builtin/providers/aws/resource_aws_emr.go

```diff
@@ -0,0 +1,413 @@
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  0,
+			},
+			"log_uri": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"applications": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"ec2_attributes": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key_name": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"subnet_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"additional_master_security_groups": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"additional_slave_security_groups": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"emr_managed_master_security_group": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"emr_managed_slave_security_group": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"instance_profile": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"bootstrap_action": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"path": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"args": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+					},
+				},
+			},
+			"tags": tagsSchema(),
+			"configurations": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"service_role": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceAwsEMRCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).emrconn
+
+	log.Printf("[DEBUG] Creating EMR cluster")
+	masterInstanceType := d.Get("master_instance_type").(string)
+	coreInstanceType := masterInstanceType
+	if v, ok := d.GetOk("core_instance_type"); ok {
+		coreInstanceType = v.(string)
+	}
+	coreInstanceCount := d.Get("core_instance_count").(int)
+
+	applications := d.Get("applications").(*schema.Set).List()
+	var userKey, subnet, extraMasterSecGrp, extraSlaveSecGrp, emrMasterSecGrp, emrSlaveSecGrp, instanceProfile, serviceRole string
+	instanceProfile = "EMR_EC2_DefaultRole"
+	ec2Attributes := d.Get("ec2_attributes").([]interface{})
+	if len(ec2Attributes) == 1 {
```

Comment:
we should probably error if they define 2+ blocks of EC2 Attributes. As is, if they have 2 blocks then nothing will happen. You can also leverage the `MaxItems` attribute of a schema, if you want to enforce that:

- https://github.com/hashicorp/terraform/blob/master/helper/schema/schema.go#L108